auth_dynamic_groups_contract/models/res_groups.py

```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+# pylint: disable=missing-docstring,no-self-use
+from odoo import api, fields, models
+
+
+class ResGroups(models.Model):
+    _inherit = 'res.groups'
+
+    is_contract_based = fields.Boolean('Based on contracts')
+
+    @api.model
+    def create(self, vals):
+        """Set dynamic_group_condition if product based."""
+        vals = self._check_contract_based(vals)
+        return super(ResGroups, self).create(vals)
+
+    @api.multi
+    def write(self, vals):
+        """Set dynamic_group_condition if product based."""
+        vals = self._check_contract_based(vals)
+        return super(ResGroups, self).create(vals)
+
+    def _check_contract_based(self, vals):
+        """Set dynamic condition for when groups are based on contract."""
+        contract_based = vals.get('is_contract_based', False)
+        # If contract based set fixed condition.
+        if contract_based:
+            vals['dynamic_group_condition'] = \
+                "user.has_group_contract(dynamic_group)"
```

Comment:
as the UI doesn't force the user to have `is_dynamic` checked, I'd propose to set this flag here unconditionally